### PR TITLE
Relax restrictions on multiple sanitizers

### PR DESCRIPTION
**Pull request by djkoloski**
Most combinations of LLVM sanitizers are legal-enough to enable simultaneously. This change will allow simultaneously enabling ASAN and shadow call stacks on supported platforms.

I used this python script to generate the mutually-exclusive sanitizer combinations:

```python
#!/usr/bin/python3

import subprocess

flags = [
    ["-fsanitize=address"],
    ["-fsanitize=leak"],
    ["-fsanitize=memory"],
    ["-fsanitize=thread"],
    ["-fsanitize=hwaddress"],
    ["-fsanitize=cfi", "-flto", "-fvisibility=hidden"],
    ["-fsanitize=memtag", "--target=aarch64-linux-android", "-march=armv8a+memtag"],
    ["-fsanitize=shadow-call-stack"],
    ["-fsanitize=kcfi", "-flto", "-fvisibility=hidden"],
    ["-fsanitize=kernel-address"],
    ["-fsanitize=safe-stack"],
    ["-fsanitize=dataflow"],
]

for i in range(len(flags)):
    for j in range(i):
        command = ["clang++"] + flags[i] + flags[j] + ["-o", "main.o", "-c", "main.cpp"]
        completed = subprocess.run(command, stderr=subprocess.DEVNULL)
        if completed.returncode != 0:
            first = flags[i][0][11:].replace('-', '').upper()
            second = flags[j][0][11:].replace('-', '').upper()
            print(f"(SanitizerSet::{first}, SanitizerSet::{second}),")
```

### CODE REVIEW

File: compiler/rustc_target/src/spec/mod.rs [hunk 1]
```diff
@@ -1316,6 +1316,32 @@ bitflags::bitflags! {
...
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::MEMORY),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::THREAD),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::HWADDRESS),
+        (SanitizerSet::KERNELADDRESS, SanitizerSet::MEMTAG),
+        (SanitizerSet::SAFESTACK, SanitizerSet::ADDRESS),
+        (SanitizerSet::SAFESTACK, SanitizerSet::LEAK),
+        (SanitizerSet::SAFESTACK, SanitizerSet::MEMORY),
+        (SanitizerSet::SAFESTACK, SanitizerSet::THREAD),
+        (SanitizerSet::SAFESTACK, SanitizerSet::HWADDRESS),
+        (SanitizerSet::SAFESTACK, SanitizerSet::KERNELADDRESS),
```
Comment (ilovepi) [inline compiler/rustc_target/src/spec/mod.rs:1342]: This doesn't quite seem right. For instance SafeStack is compatible with all sanitizers. In fact, Fuchsia uses it everywhere on X86_64 w/ ASAN. I can't recall if we fully support TSAN or KASAN yet, but it shouldn't be incompatible. 

The canonical definition in clang is here https://github.com/llvm/llvm-project/blob/70b79a9ccd03f93fc4c8464a91b6bef3aab322d3/clang/lib/Driver/SanitizerArgs.cpp#L520

Comment (djkoloski) [reply compiler/rustc_target/src/spec/mod.rs:1342]: The definitions you linked seem to suggest that SafeStack is not compatible with all sanitizers:

```
std::make_pair(SanitizerKind::SafeStack,
               (TC.getTriple().isOSFuchsia() ? SanitizerMask()
                                             : SanitizerKind::Leak) |
                   SanitizerKind::Address | SanitizerKind::HWAddress |
                   SanitizerKind::Thread | SanitizerKind::Memory |
                   SanitizerKind::KernelAddress),
```

Though this change doesn't currently capture that Fuchsia specifically allows LSAN + SafeStack.

Comment (ilovepi) [reply compiler/rustc_target/src/spec/mod.rs:1342]: Hmm, maybe it is ok, I realize now that we disable SafeStack w/ ASAN, so disregard. But the link can be a good cross reference.

**Comment by ilovepi** [comment]
> Since Address Sanitizer includes the leak detection, combining it with Leak Sanitizer should omit the Leak Sanitizer runtime library:
> 
> https://github.com/rust-lang/llvm-project/blob/5399a24c66cb6164cf32280e7d300488c90d5765/clang/include/clang/Driver/SanitizerArgs.h#L92-L96

That's a good point. I had assumed `rustc` already had similar logic, but I suppose if multiple sanitizers were disallowed, maybe it did not. This is some more evidence that on the LLVM side, we should probably keep that logic in a support library that frontends like `clang` and `rustc` can reuse without duplicating the same logic everywhere.

File: compiler/rustc_target/src/spec/mod.rs [hunk 2]
```diff
@@ -1316,6 +1316,32 @@ bitflags::bitflags! {
 rustc_data_structures::external_bitflags_debug! { SanitizerSet }
 
 impl SanitizerSet {
+    const MUTUALLY_EXCLUSIVE: &'static [(SanitizerSet, SanitizerSet)] = &[
```
Comment (tmandry) [inline compiler/rustc_target/src/spec/mod.rs:1319]: Can you add the link @ilovepi posted as a cross reference here?

**Comment by rcvalle** [comment]
LGTM (@cuviper FYI). Thank you for your time and for working on this, @djkoloski! Much appreciated.

File: compiler/rustc_target/src/spec/mod.rs (hunk 2 above)
Comment (djkoloski) [reply compiler/rustc_target/src/spec/mod.rs:1321]: Done

**Comment by cuviper** [comment]
Thanks!

@bors r=cuviper,rcvalle

**Pull request merged**